### DISCUSSION TITLE
Best way to Format/Transform Form Field Values While Changing?

### DISCUSSION THREAD

**Discussion by elibroftw** [mantinedev/mantine#7018] [Q&A]
I have a TextInput that needs to format a long string of characters like so: XXXX - XXXX - XXX. So if they user types in XXXXX, I need to show it like XXXX - X. I have the function to do this transformation, but the biggest headache right now is updating the form values. What is the best practice? I'm going to use form actions. I also can't use uncontrolled this way.

**Comment by arvind0598** [comment]
I guess there's two ways to go about this if you wanna use uncontrolled mode:
- There's an example [here](https://mantine.dev/core/input/#change-input-element) for masked inputs. I think you'll have to work to support uncontrolled mode for custom inputs though ([here](https://mantine.dev/form/uncontrolled/#uncontrolled-mode-in-custom-components)).
- You could possibly use a regular text input and use [onValuesChange](https://mantine.dev/form/uncontrolled/#uncontrolled-mode-in-custom-components) to monitor and format the value?
